### PR TITLE
ddl: support multi-schema change for modify columns

### LINKED ISSUE #39: Support multi-schema change for `modify/change column`

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #39

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: ddl/column.go [hunk 1]
```diff
@@ -440,11 +440,13 @@ type modifyingColInfo struct {
...
 	changingIdxs          []*model.IndexInfo
 	pos                   *ast.ColumnPosition
+	removedIdxs           []int64
 }
 
 func getModifyColumnInfo(t *meta.Meta, job *model.Job) (*model.DBInfo, *model.TableInfo, *model.ColumnInfo, *modifyingColInfo, error) {
 	modifyInfo := &modifyingColInfo{pos: &ast.ColumnPosition{}}
-	err := job.DecodeArgs(&modifyInfo.newCol, &modifyInfo.oldColName, modifyInfo.pos, &modifyInfo.modifyColumnTp, &modifyInfo.updatedAutoRandomBits, &modifyInfo.changingCol, &modifyInfo.changingIdxs)
+	err := job.DecodeArgs(&modifyInfo.newCol, &modifyInfo.oldColName, modifyInfo.pos, &modifyInfo.modifyColumnTp,
+		&modifyInfo.updatedAutoRandomBits, &modifyInfo.removedIdxs, &modifyInfo.changingCol, &modifyInfo.changingIdxs)
```
Comment (Defined2014) [inline ddl/column.go]: Does it affect upgrade?

File: ddl/column.go [hunk 2]
```diff
@@ -796,32 +870,49 @@ func adjustTableInfoAfterModifyColumnWithData(tblInfo *model.TableInfo, pos *ast
...
+		replaceOldIndexes(tblInfo, indexesToRemove)
 	}
 	// Move the new column to a correct offset.
 	destOffset, err := locateOffsetToMove(changingCol.Offset, pos, tblInfo)
 	if err != nil {
 		return errors.Trace(err)
 	}
 	tblInfo.MoveColumnInfo(changingCol.Offset, destOffset)
+
+	checkOffsets(tblInfo)
```
Comment (Defined2014) [inline ddl/column.go]: delete it

File: ddl/column.go [hunk 3]
```diff
@@ -856,16 +947,29 @@ func replaceOldIndexes(tblInfo *model.TableInfo, changingIdxs []*model.IndexInfo
...
+		var hasChangingCol bool
 		for i, col := range idx.Columns {
 			if col.Name.L == oldName.L {
 				idx.Columns[i].Name = newName
 				idx.Columns[i].Offset = newOffset
-				break
+			} else {
+				if !hasChangingCol {
+					hasChangingCol = len(getChangingColumnOriginName(col)) != len(col.Name.L)
+				}
```
Comment (Defined2014) [inline ddl/column.go]:
```suggestion
				if getChangingColumnOriginName(col) != col.Name.L {
									hasChangingCol = true
									break
				}
```

And one question here, why only compare `len`?

File: ddl/column.go [hunk 4]
```diff
@@ -856,16 +947,29 @@ func replaceOldIndexes(tblInfo *model.TableInfo, changingIdxs []*model.IndexInfo
...
 
-func updateNewIndexesCols(tblInfo *model.TableInfo, oldName, newName model.CIStr, newOffset int) {
-	for _, idx := range tblInfo.Indices {
+// updateNewIndexesCols updates the changing indexes column name&offset, and
+// filter out the indexes that can be removed.
+func updateNewIndexesCols(changingIdxs []*model.IndexInfo,
+	oldName, newName model.CIStr, newOffset int) []*model.IndexInfo {
+	indexesToRemove := make([]*model.IndexInfo, 0, len(changingIdxs))
+	for _, idx := range changingIdxs {
+		var hasChangingCol bool
```
Comment (Defined2014) [inline ddl/column.go]: `hasOtherChangingCols` is better

File: ddl/column.go [hunk 5]
```diff
@@ -880,11 +984,24 @@ func updateChangingCol(col *model.ColumnInfo, newName model.CIStr, newOffset int
...
+			if tblInfo.Columns[idxCol.Offset].ID == oldColID {
+				indexInfos = append(indexInfos, idx)
+				break
+			}
+		}
+	}
+	return indexInfos
+}
+
+func getRelatedIndexIDs(tblInfo *model.TableInfo, colID int64) []int64 {
```
Comment (Defined2014) [inline ddl/column.go:976]: Two funcs `getRelatedIndexIDs` and `getChangingIndexInfos` are very similar.
Use similar name is better, like `getRelatedIndexXXX` or `getChangingIndexXXX`.
Or combine them, return two slices.

File: ddl/column.go [hunk 6]
```diff
@@ -549,49 +551,69 @@ func (w *worker) onModifyColumn(d *ddlCtx, t *meta.Meta, job *model.Job) (ver in
...
-			modifyInfo.changingIdxs = append(modifyInfo.changingIdxs, newIdxInfo)
 		}
-		tblInfo.Indices = append(tblInfo.Indices, modifyInfo.changingIdxs...)
+		job.Args = append(job.Args, indexesToRemove)
 	} else {
-		tblInfo.Columns[len(tblInfo.Columns)-1] = modifyInfo.changingCol
-		copy(tblInfo.Indices[len(tblInfo.Indices)-len(modifyInfo.changingIdxs):], modifyInfo.changingIdxs)
+		changingCol = model.FindColumnInfoByID(tblInfo.Columns, modifyInfo.changingCol.ID)
+	}
+	if changingCol == nil {
```
Comment (Defined2014) [inline ddl/column.go]: Move into else?

File: ddl/index.go [hunk 7]
```diff
@@ -1509,17 +1516,43 @@ func (w *worker) updateReorgInfoForPartitions(t table.PartitionedTable, reorg *r
...
-	idxInfos := make([]*model.IndexInfo, 0, len(indexes))
-	offsets := make([]int, 0, len(indexes))
-	for _, idxInfo := range indexes {
-		for i, c := range idxInfo.Columns {
-			if strings.EqualFold(colName, c.Name.L) {
-				idxInfos = append(idxInfos, idxInfo)
-				offsets = append(offsets, i)
+type indexesToChange struct {
+	indexInfo  *model.IndexInfo
+	isChanging bool // whether the index is created by modify column
```
Comment (Defined2014) [inline ddl/index.go]: Name and meaning are not very appropriate.

File: ddl/multi_schema_change_test.go [hunk 8]
```diff
@@ -675,10 +675,64 @@ func TestMultiSchemaChangeModifyColumns(t *testing.T) {
...
 	tk.MustExec("insert into t values (9223372036854775807, 9223372036854775807, 1);")
 	tk.MustQuery("select * from t;").Check(
-		testkit.Rows("9223372036854775807 9223372036854775807 1"))
+		testkit.Rows("1 2 3", "9223372036854775807 9223372036854775807 1"))
+
+	// Modify index-covered columns.
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t (a int, b int, c int, index i1(a), index i2(b), index i3(c), index i4(a, b), index i5(a, b, c));")
+	tk.MustExec("insert into t values (1, 2, 3);")
+	tk.MustExec("alter table t modify column a tinyint, modify column b tinyint, modify column c tinyint;")
```
Comment (Defined2014) [inline ddl/multi_schema_change_test.go:691]: add check about delete range for tmp idxes.

File: ddl/column.go (hunk 3 above)
Comment (tangenta) [reply ddl/column.go]: Different length means it is a changing column.

File: ddl/index.go (hunk 7 above)
Comment (tangenta) [reply ddl/index.go]: Do you have any suggestion?

File: ddl/column.go (hunk 3 above)
Comment (Defined2014) [reply ddl/column.go]: I know that, but it looks a little bit strange. I think compare name directly is better.

File: ddl/index.go (hunk 7 above)
Comment (Defined2014) [reply ddl/index.go]: How about `isGenerated`?

Comment (tangenta) [reply ddl/index.go]: It seems easy to mix this concept up with "generated columns"...

We just need to mention "the index is in a middle(invisible) state".

File: ddl/column.go (hunk 3 above)
Comment (tangenta) [reply ddl/column.go]: OK.

**Review by Defined2014 — APPROVED** [review]

**Pull request merged**